lib/internal/process/promises.js

```diff
@@ -18,32 +20,51 @@ function setupPromises(scheduleMicrotasks) {
 
   function unhandledRejection(promise, reason) {
     hasBeenNotifiedProperty.set(promise, false);
+    promiseToGuidProperty.set(promise, lastPromiseId++);
     addPendingUnhandledRejection(promise, reason);
   }
 
   function rejectionHandled(promise) {
     var hasBeenNotified = hasBeenNotifiedProperty.get(promise);
     if (hasBeenNotified !== undefined) {
       hasBeenNotifiedProperty.delete(promise);
+      const uid = promiseToGuidProperty.get(promise);
+      promiseToGuidProperty.delete(promise);
       if (hasBeenNotified === true) {
         process.nextTick(function() {
-          process.emit('rejectionHandled', promise);
+          if (!process.emit('rejectionHandled', promise)) {
+            process.emitWarning(
+              'Promise rejection was handled asynchronously ' +
+              `(rejection id: ${uid}) `,
+              'PromiseRejectionHandledWarning');
+          }
         });
       }
 
     }
   }
 
+  var deprecationWarned = false;
   function emitPendingUnhandledRejections() {
-    var hadListeners = false;
+    let hadListeners = false;
     while (pendingUnhandledRejections.length > 0) {
-      var promise = pendingUnhandledRejections.shift();
-      var reason = pendingUnhandledRejections.shift();
+      const promise = pendingUnhandledRejections.shift();
+      const reason = pendingUnhandledRejections.shift();
       if (hasBeenNotifiedProperty.get(promise) === false) {
         hasBeenNotifiedProperty.set(promise, true);
+        const uid = promiseToGuidProperty.get(promise);
         if (!process.emit('unhandledRejection', reason, promise)) {
-          // Nobody is listening.
-          // TODO(petkaantonov) Take some default action, see #830
+          process.emitWarning('Unhandled promise rejection ' +
+                              `(rejection id: ${uid}): ${reason}`,
+                              'UnhandledPromiseRejectionWarning');
+          if (!deprecationWarned) {
+            deprecationWarned = true;
+            process.emitWarning(
+              'Unhandled promise rejections are deprecated. In the future, ' +
+              'an error will be thrown when a promise rejection is not ' +
```

Comment:
OK, I'm back up to date. The conversation about "when" the error crashes — whether on GC, or immediately — should probably be left for later, to see how the community at large reacts to the warning messages. If, by this time next year, there's still a lot of folks leaning on handling previously unhandled rejections, we can lean towards crashing on GC, otherwise we can crash immediately. To that end, a slight spin on @jasnell's wording:

```
In the future, promise rejections that are not handled will
terminate the Node.js process with a non-zero exit code.
```

(Without a specific "when" — whether at GC, or at unhandled-rejection-time)